Source/Editor/DualityEditor/Forms/ObjectRefSelectionDialog.cs

```diff
@@ -0,0 +1,281 @@
+
+			public IContentRef ResourceReference
+			{
+				get { return this._resourceReference; }
+				set { this._resourceReference = value; }
+			}
+
+			private GameObject _gameObjectReference;
+
+			public GameObject GameObjectReference
+			{
+				get { return this._gameObjectReference; }
+				set { this._gameObjectReference = value; }
+			}
+
+			private Component _componentReference;
+
+			public Component ComponentReference
+			{
+				get { return this._componentReference; }
+				set { this._componentReference = value; }
+			}
+		}
+
+		public Color PathColor { get; set; }
+		public Type FilteredType { get; set; }
+
+		public string ResourcePath { get; set; }
+
+		public IContentRef ResourceReference { get; set; }
+		public GameObject GameObjectReference { get; set; }
+		public Component ComponentReference { get; set; }
+
+		private TreeModel Model { get; set; }
+
+		public ObjectRefSelectionDialog()
+		{
+			InitializeComponent();
+
+			this.Model = new TreeModel();
+
+			this.PathColor = Color.Gray;
+
+			this.objectReferenceListing.SelectionChanged += this.ResourceListingOnSelectedIndexChanged;
+			this.objectReferenceListing.DoubleClick += this.ResourceListingOnDoubleClick;
+
+			this.txtFilterInput.KeyDown += this.TxtFilterInputOnKeyDown;
+
+			this.nodeName.DrawText += this.NodeName_OnDrawText;
+			this.nodePath.DrawText += this.NodePath_OnDrawText;
+		}
+
+		private void TxtFilterInputOnKeyDown(object sender, KeyEventArgs keyEventArgs)
+		{
+			TreeNodeAdv tmp = null;
+
+			if (keyEventArgs.KeyCode == Keys.Down)
+			{
+				tmp = this.objectReferenceListing.SelectedNode.NextNode;
+			}
+			else if (keyEventArgs.KeyCode == Keys.Up)
+			{
+				tmp = this.objectReferenceListing.SelectedNode.PreviousNode;
+			}
+
+			if (tmp != null)
+			{
+				this.objectReferenceListing.SelectedNode = tmp;
+			}
+		}
+
+		private void NodePath_OnDrawText(object sender, DrawTextEventArgs drawTextEventArgs)
+		{
+			ReferenceNode node = drawTextEventArgs.Node.Tag as ReferenceNode;
+
+			if (node == null)
+			{
+				return;
+			}
+
+			drawTextEventArgs.TextColor = this.PathColor;
+		}
+
+		private void NodeName_OnDrawText(object sender, DrawTextEventArgs drawTextEventArgs)
+		{
+			ReferenceNode node = drawTextEventArgs.Node.Tag as ReferenceNode;
+
+			if (node == null)
+			{
+				return;
+			}
+
+			drawTextEventArgs.TextColor = this.objectReferenceListing.ForeColor;
+		}
+
+		protected override void OnShown(EventArgs e)
+		{
+			base.OnShown(e);
+
+			this.objectReferenceListing.ClearSelection();
+			this.objectReferenceListing.Model = this.Model;
+
+			this.objectReferenceListing.BeginUpdate();
+			this.Model.Nodes.Clear();
+			
+			if (this.FilteredType.IsSubclassOf(typeof(GameObject)) || this.FilteredType == typeof(GameObject))
+			{
+				this.Text = "Select a GameObject";
+
+				foreach (GameObject currentObject in Scene.Current.AllObjects)
+				{
+					ReferenceNode tmpNode = new ReferenceNode(currentObject);
+
+					this.Model.Nodes.Add(tmpNode);

```

Comment:
Suggestion: The newline between the two lines of the inner loop doesn't seem to be necessary. Keep it if you want (not a change request), but as a general rule of thumb, I try to make every newline express something meaningful and avoid those where it's not clear what exactly they're separating, like logical code blocks. Here, instantiating the node and adding the node each are so short and self-documenting that they might as well be regarded as a single logical unit, in which case they don't really need the newline.